File: api/HabitLogic.cs
public static class HabitLogic
{
    public static Habit CreateHabit(string name, int dailyTarget, string? color)
    {
        return new Habit
        {
            Id = Guid.NewGuid().ToString(),
            Name = name,
            DailyTarget = Math.Clamp(dailyTarget, 1, 10),
            Color = color,
            CreatedAt = DateTime.UtcNow.ToString("o"),
            CurrentStreak = 0,
            LongestStreak = 0,
            TotalCompletions = 0,
            DayRecords = new Dictionary<string, HabitDayRecord>()
        };
    }

    private static HabitDayRecord EnsureDayRecord(Habit habit, string date)
    {
        if (!habit.DayRecords.ContainsKey(date))
        {
            habit.DayRecords[date] = new HabitDayRecord
            {
                Date = date,
                Count = 0,
                Completed = false
            };
        }
        return habit.DayRecords[date];
    }

    private static bool IsCompleteOnDate(Habit habit, string date)
    {
        return habit.DayRecords.TryGetValue(date, out var record) && record.Completed;
    }

    public static bool ProcessMissedDays(Habit habit, string today)
    {
        if (habit.LastEvaluatedDate == null)
        {
            habit.LastEvaluatedDate = today;
            return false;
        }

        if (habit.LastEvaluatedDate == today)
            return false;

        var date = DateUtils.GetNextDateString(habit.LastEvaluatedDate);
        var missedAny = false;

        while (string.Compare(date, today, StringComparison.Ordinal) < 0)
        {
            if (!habit.DayRecords.TryGetValue(date, out var record) || !record.Completed)
                missedAny = true;
            date = DateUtils.GetNextDateString(date);
        }

        if (missedAny)
            habit.CurrentStreak = 0;

        habit.LastEvaluatedDate = today;
        return missedAny;
    }

    public static void RecalculateStats(Habit habit, string today)
    {
        // Total completions
        habit.TotalCompletions = habit.DayRecords.Values.Count(r => r.Completed);

        // Current streak
        var streak = 0;
        var anchor = today;
        if (!IsCompleteOnDate(habit, today))
            anchor = DateUtils.GetYesterdayDateString(today);
        while (IsCompleteOnDate(habit, anchor))
        {
            streak++;
            anchor = DateUtils.GetYesterdayDateString(anchor);
        }
        habit.CurrentStreak = streak;

        // Longest streak
        var dates = habit.DayRecords.Keys.OrderBy(d => d).ToList();
        var longest = 0;
        var running = 0;
        string? prevCompleted = null;
        foreach (var d in dates)
        {
            if (!habit.DayRecords[d].Completed)
            {
                running = 0;
                prevCompleted = null;
                continue;
            }
            if (prevCompleted != null && d == DateUtils.GetNextDateString(prevCompleted))
                running++;
            else
                running = 1;
            if (running > longest)
                longest = running;
            prevCompleted = d;
        }
        habit.LongestStreak = longest;
    }

    public static void Increment(Habit habit, string today)
    {
        ProcessMissedDays(habit, today);
        var record = EnsureDayRecord(habit, today);
        if (record.Count < habit.DailyTarget)
            record.Count++;
        record.Completed = record.Count >= habit.DailyTarget;
        RecalculateStats(habit, today);
    }

    public static void Undo(Habit habit, string today)
    {
        ProcessMissedDays(habit, today);
        var record = EnsureDayRecord(habit, today);
        if (record.Count > 0)
            record.Count--;
        record.Completed = record.Count >= habit.DailyTarget;
        RecalculateStats(habit, today);
    }

    public static void SetDayCount(Habit habit, string date, int count, string today)
    {
        var clamped = Math.Clamp(count, 0, habit.DailyTarget);
        if (clamped == 0)
        {
            habit.DayRecords.Remove(date);
        }
        else
        {
            if (!habit.DayRecords.ContainsKey(date))
                habit.DayRecords[date] = new HabitDayRecord { Date = date, Count = 0, Completed = false };
            habit.DayRecords[date].Count = clamped;
            habit.DayRecords[date].Completed = clamped >= habit.DailyTarget;
        }
        RecalculateStats(habit, today);
    }

    public static void Edit(Habit habit, string name, int dailyTarget, string? color, string today)
    {
        habit.Name = name;
        habit.DailyTarget = Math.Clamp(dailyTarget, 1, 10);
        habit.Color = color;
        foreach (var key in habit.DayRecords.Keys.ToList())
        {
            habit.DayRecords[key].Completed = habit.DayRecords[key].Count >= habit.DailyTarget;
        }
        RecalculateStats(habit, today);
    }
}

File: src/domain/habits/habit.logic.ts
import type { Habit, HabitDayRecord, DateString } from './habit.types';
import { getYesterdayDateString, getNextDateString } from '../../utils/dates';

export function createHabit(name: string, dailyTarget: number): Habit {
  return {
    id: crypto.randomUUID(),
    name,
    dailyTarget: Math.max(1, Math.min(10, dailyTarget)),
    createdAt: new Date().toISOString(),
    currentStreak: 0,
    longestStreak: 0,
    totalCompletions: 0,
    dayRecords: {},
  };
}

function ensureDayRecord(habit: Habit, date: DateString): HabitDayRecord {
  if (!habit.dayRecords[date]) {
    habit.dayRecords[date] = {
      date,
      count: 0,
      completed: false,
    };
  }
  return habit.dayRecords[date];
}

function isHabitCompleteOnDate(habit: Habit, date: DateString): boolean {
  const record = habit.dayRecords[date];
  return !!record && record.completed;
}

export function processMissedDays(habit: Habit, today: DateString): boolean {
  if (!habit.lastEvaluatedDate) {
    habit.lastEvaluatedDate = today;
    return false;
  }

  if (habit.lastEvaluatedDate === today) {
    return false;
  }

  let date = getNextDateString(habit.lastEvaluatedDate);
  let missedAnyDay = false;

  while (date < today) {
    const record = habit.dayRecords[date];
    if (!record || !record.completed) {
      missedAnyDay = true;
    }
    date = getNextDateString(date);
  }

  if (missedAnyDay) {
    habit.currentStreak = 0;
  }

  habit.lastEvaluatedDate = today;
  return missedAnyDay;
}

function recalculateTotalCompletions(habit: Habit): void {
  let total = 0;
  for (const record of Object.values(habit.dayRecords)) {
    if (record.completed) {
      total += 1;
    }
  }
  habit.totalCompletions = total;
}

function recalculateCurrentStreak(habit: Habit, today: DateString): void {
  let streak = 0;
  let anchorDate = today;

  if (!isHabitCompleteOnDate(habit, today)) {
    anchorDate = getYesterdayDateString(today);
  }

  while (isHabitCompleteOnDate(habit, anchorDate)) {
    streak += 1;
    anchorDate = getYesterdayDateString(anchorDate);
  }

  habit.currentStreak = streak;
}

function recalculateLongestStreak(habit: Habit): void {
  const dates = Object.keys(habit.dayRecords).sort();

  let longest = 0;
  let running = 0;
  let previousCompletedDate: string | undefined;

  for (const date of dates) {
    const record = habit.dayRecords[date];

    if (!record.completed) {
      running = 0;
      previousCompletedDate = undefined;
      continue;
    }

    if (
      previousCompletedDate &&
      date === getNextDateString(previousCompletedDate)
    ) {
      running += 1;
    } else {
      running = 1;
    }

    if (running > longest) {
      longest = running;
    }

    previousCompletedDate = date;
  }

  habit.longestStreak = longest;
}

export function recalculateHabitStats(habit: Habit, today: DateString): void {
  recalculateTotalCompletions(habit);
  recalculateCurrentStreak(habit, today);
  recalculateLongestStreak(habit);
}

export type IncrementResult = {
  wasCompleted: boolean;
  newCount: number;
  dailyTarget: number;
};

export function incrementHabit(habit: Habit, today: DateString): IncrementResult {
  processMissedDays(habit, today);

  const record = ensureDayRecord(habit, today);

  if (record.count < habit.dailyTarget) {
    record.count += 1;
  }

  record.completed = record.count >= habit.dailyTarget;

  recalculateHabitStats(habit, today);

  return {
    wasCompleted: record.completed,
    newCount: record.count,
    dailyTarget: habit.dailyTarget,
  };
}

export type UndoResult = {
  wasCompleted: boolean;
  lostCompletion: boolean;
  newCount: number;
  dailyTarget: number;
};

export function undoHabit(habit: Habit, today: DateString): UndoResult {
  processMissedDays(habit, today);

  const record = ensureDayRecord(habit, today);
  const wasCompletedBefore = record.completed;

  if (record.count > 0) {
    record.count -= 1;
  }

  record.completed = record.count >= habit.dailyTarget;

  recalculateHabitStats(habit, today);

  return {
    wasCompleted: record.completed,
    lostCompletion: wasCompletedBefore && !record.completed,
    newCount: record.count,
    dailyTarget: habit.dailyTarget,
  };
}

export function getCompletionRate(habit: Habit): number {
  const records = Object.values(habit.dayRecords);
  if (records.length === 0) return 0;
  const completed = records.filter((r) => r.completed).length;
  return Math.round((completed / records.length) * 100);
}

export type DayStatus = 'complete' | 'in-progress' | 'missed' | 'today-empty';

export function getDayStatus(
  habit: Habit,
  date: DateString,
  today: DateString
): DayStatus {
  const record = habit.dayRecords[date];

  if (record && record.completed) return 'complete';
  if (date === today && record && record.count > 0) return 'in-progress';
  if (date === today) return 'today-empty';
  return 'missed';
}
